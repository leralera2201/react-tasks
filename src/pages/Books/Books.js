import React, { Component } from "react";
import { connect } from "react-redux";

import { Loader, BookCard, BooksPagination, Error } from "../../components";

import {fetchBooksStart, paginateBooks} from "./actions/books.actions";
import {
  booksFetchErrorSelector,
  booksFetchIsStatusInProgress,
  booksFetchItemsSelector,
  booksFetchCurrentPageSelector,
  booksFetchBooksPerPagePageSelector,
  booksFetchCurrentItemsSelector
} from "./selectors/books.selectors";
import {Button} from "reactstrap";

class Books extends Component {
  componentDidMount() {
    const { getAllBooks, paginate, location: {state} } = this.props;
    if(state && state.withoutFetch) {
      paginate()
      return
    }
    getAllBooks();
  }

  createBook = () => {
    const { history } = this.props;
    history.push('/books/create')
  }

  render() {
    const {
      loading,
      currentBooks,
      error,
      booksPerPage,
      currentPage,
      books,
      paginate,
    } = this.props;
    const booksLength = books.length;

    const pages = [];

    const totalPages = Math.ceil(booksLength / booksPerPage);
    for (let i = 1; i <= totalPages; i++) {
      pages.push(i);
    }

    return (
      <div className="pt-5">
        <div className="container">
          <Button onClick={this.createBook}>
            Create book
          </Button>
        {loading && <Loader />}
        {!loading && !error && currentBooks.length > 0 && (
            <>
              <div className="flex-space-between">
                {currentBooks.map((book) => (
                  <BookCard book={book} key={book.id} />
                ))}
              </div>
              <BooksPagination
                currentPage={currentPage}
                onPaginate={paginate}
                pages={pages}
              />
            </>
        )}
        {!error && !loading && booksLength === 0 && <div>No books</div>}
        {error && <Error error={error} />}
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  loading: booksFetchIsStatusInProgress(state),
  books: booksFetchItemsSelector(state),
  error: booksFetchErrorSelector(state),
  currentPage: booksFetchCurrentPageSelector(state),
  booksPerPage: booksFetchBooksPerPagePageSelector(state),
  currentBooks: booksFetchCurrentItemsSelector(state),
});

const mapDispatchToProps = (dispatch) => ({
  getAllBooks: () => dispatch(fetchBooksStart()),
  paginate: (page) => dispatch(paginateBooks(page)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Books);
