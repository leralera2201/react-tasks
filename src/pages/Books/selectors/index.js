import { createSelector } from "reselect";

const booksSelector = (state) => state.books;

export const booksDataSelector = createSelector(
  booksSelector,
  ({ data }) => data
);

export const booksLoadingSelector = createSelector(
  booksSelector,
  ({ loading }) => loading
);

export const booksErrorSelector = createSelector(
  booksSelector,
  ({ error }) => error
);

export const bookItemsSelector = createSelector(
  booksDataSelector,
  ({ books }) => books
);

export const currentBookItemsSelector = createSelector(
  booksDataSelector,
  ({ currentBooks }) => currentBooks
);

export const currentPageSelector = createSelector(
  booksDataSelector,
  ({ currentPage }) => currentPage
);

export const booksPerPageSelector = createSelector(
  booksDataSelector,
  ({ booksPerPage }) => booksPerPage
);
