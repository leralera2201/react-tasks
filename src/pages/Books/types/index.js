export const BOOKS_REQUEST = "BOOKS_REQUEST";
export const SET_BOOKS = "SET_BOOKS";
export const PAGINATE_BOOKS = "PAGINATE_BOOKS";
export const BOOKS_ERROR = "BOOKS_ERROR";
