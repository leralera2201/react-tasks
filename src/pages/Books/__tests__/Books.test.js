import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Books } from '../Books';

describe('Books', () => {

  const props = {
    loading: true,
    currentBooks: [],
    error: null,
    booksPerPage: 20,
    currentPage: 1,
    books: [],
    paginate: jest.fn(),
    getAllBooks: jest.fn(),
    location: {
      state: {},
    },
    history: { push: jest.fn(), location: {}, listen: jest.fn() },
  };

  it('render with loading', () => {
    const wrapper = shallow(<Books {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('render no books', () => {
    props.loading = false;
    const wrapper = shallow(<Books {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('render books', () => {
    props.currentBooks = [
      {
        title: 'd',
        description: 'd',
        id: 'd',
      },
    ];
    const wrapper = shallow(<Books {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('history change after create click', () => {
    const wrapper = shallow(<Books {...props} />);
    const button = wrapper.find('#create-button').first();
    button.simulate('click');
    expect(props.history.push.mock.calls[0][0]).toEqual('/books/create');
  });


  it('without fetch render', () => {
    props.location.state = { withoutFetch: true };
    shallow(<Books {...props} />);

    expect(props.getAllBooks).toBeCalledTimes(0);
  });

  it('render with error', () => {
    props.error = 'Some error';
    const wrapper = shallow(<Books {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
