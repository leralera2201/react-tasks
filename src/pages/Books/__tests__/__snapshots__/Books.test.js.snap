// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books render books 1`] = `
<div
  className="pt-5"
>
  <div
    className="container"
  >
    <Button
      color="secondary"
      id="create-button"
      onClick={[Function]}
      tag="button"
    >
      Create book
    </Button>
    <div
      className="flex-space-between"
    >
      <BookCard
        book={
          Object {
            "description": "d",
            "id": "d",
            "title": "d",
          }
        }
        key="d"
      />
    </div>
    <BooksPagination
      currentPage={1}
      onPaginate={[MockFunction]}
      pages={Array []}
    />
    <div
      id="no-books"
    >
      No books
    </div>
  </div>
</div>
`;

exports[`Books render no books 1`] = `
<div
  className="pt-5"
>
  <div
    className="container"
  >
    <Button
      color="secondary"
      id="create-button"
      onClick={[Function]}
      tag="button"
    >
      Create book
    </Button>
    <div
      id="no-books"
    >
      No books
    </div>
  </div>
</div>
`;

exports[`Books render with error 1`] = `
<div
  className="pt-5"
>
  <div
    className="container"
  >
    <Button
      color="secondary"
      id="create-button"
      onClick={[Function]}
      tag="button"
    >
      Create book
    </Button>
    <Error
      error="Some error"
    />
  </div>
</div>
`;

exports[`Books render with loading 1`] = `
<div
  className="pt-5"
>
  <div
    className="container"
  >
    <Button
      color="secondary"
      id="create-button"
      onClick={[Function]}
      tag="button"
    >
      Create book
    </Button>
    <Loader />
  </div>
</div>
`;
